custom/icds_reports/static/js/directives/download/download.directive.js

```diff
@@ -229,107 +204,19 @@ function DownloadController($rootScope, $location, locationHierarchy, locationsS
-                        });
-                        if (vm.preventShowingAllOption(sortedLocations)) {
-                            locationsCache[parentId] = sortedLocations;
-                        } else if (selectedLocation.user_have_access) {
-                            locationsCache[parentId] = [ALL_OPTION].concat(sortedLocations);
-                        } else {
-                            locationsCache[parentId] = sortedLocations;
-                        }
-                    }
-                }
-
-                initHierarchy();
-
-                var levelOfSelectedLocation = _.findIndex(vm.hierarchy, function(locationTypes) {
-                    return _.contains(locationTypes.map(function(x) { return x.name; }), selectedLocation.location_type_name);
-                });
-                vm.selectedLocations[levelOfSelectedLocation] = vm.selectedLocationId;
-                vm.onSelect(selectedLocation, levelOfSelectedLocation);
-
-                levelOfSelectedLocation -= 1;
-
-                while(levelOfSelectedLocation >= 0) {
-                    var childSelectedId = vm.selectedLocations[levelOfSelectedLocation + 1];
-                    var childSelected = _.find(locations, function(location) {
-                        return location.location_id === childSelectedId;
-                    });
-                    vm.selectedLocations[levelOfSelectedLocation] = childSelected.parent_id;
-                    levelOfSelectedLocation -= 1;
-                }
-
-                var levels = [];
-                window.angular.forEach(vm.levels, function (value) {
-                    if (value.id > selectedLocationIndex()) {
-                        levels.push(value);
-                    }
-                });
-                vm.groupByLevels = levels;
-                vm.selectedLevel = selectedLocationIndex() + 1;
-            });
-        } else {
-            initHierarchy();
-            vm.myPromise = locationsService.getRootLocations().then(function(data) {
-                locationsCache.root = [NATIONAL_OPTION].concat(data.locations);
-            });
-            vm.groupByLevels = vm.levels;
-        }
+        initHierarchy();
+        locationsCache = locationsService.initLocations(vm, locationsCache);
     };
 
     init();
 
     vm.getPlaceholder = function(locationTypes) {
-        return _.map(locationTypes, function(locationType) {
-            if (locationType.name === 'state') {
-                if (vm.isChildBeneficiaryListSelected()) {
-                    return 'Select State';
-                } else {
-                    return NATIONAL_OPTION.name;
-                }
-            }
-            return locationType.name;
-        }).join(', ');
-    };
-
-    vm.showErrorMessage = function () {
-        return vm.selectedIndicator === 6 && selectedLocationIndex() !== 4;
+        return locationsService.getLocationPlaceholder(locationTypes, vm.isChildBeneficiaryListSelected())
```

Comment:
nit: this would probably be more readable and future-proof if you defined `vm.allowNationalSelection = vm.isChildBeneficiaryListSelected` just in case that logic changes in the future (kind of like you did with the variable below but just sticking it on the controller)